src/relax/ir/expr_functor.cc

```
@@ -580,18 +597,14 @@ void ExprMutator::VisitBinding_(const MatchShapeNode* binding) {
 
   Var new_var;
   if (binding->var.defined()) {
-    // in the case of `x = R.match_shape(val, pattern)`, we want `x` to directly get `pattern` as
-    // the shape when `val` is a tensor.
-    Optional<Expr> new_shape;
-    Type new_type = new_value->checked_type_;
-    if (new_value->checked_type_.defined() && new_value->checked_type_.as<DynTensorTypeNode>()) {
-      new_shape = new_pattern;
-      ICHECK(new_shape->IsInstance<ShapeExprNode>());
-      int ndim = Downcast<ShapeExpr>(new_shape.value())->values.size();
-      new_type = DynTensorType(ndim, new_value->checked_type_.as<DynTensorTypeNode>()->dtype);
+    StructInfo new_sinfo = GetStructInfo(new_value);
+
+    if (auto* ptr = new_sinfo.as<TensorStructInfoNode>()) {
+      new_sinfo = TensorStructInfo(new_pattern, ptr->dtype);
     }
```

Shall we throw an error if `new_sinfo` is not a TensorStructInfo?

---

MatchShape had an overloaded behavior that also takes shape. We will migrate to MatchCast later, which do not have such behavior